futures-util/src/task/arc_wake.rs

```diff
@@ -32,39 +23,43 @@ pub trait ArcWake {
     ///
     /// If `wake()` is called on the returned `Waker`,
     /// the `wake()` function that is defined inside this trait will get called.
-    fn into_waker(wake: Arc<Self>) -> Waker where Self: Sized
+    fn into_waker(self: Arc<Self>) -> Waker where Self: Sized
     {
-        let ptr = Arc::into_raw(wake) as *const();
+        let ptr = Arc::into_raw(self) as *const();
 
         unsafe {
             Waker::new_unchecked(RawWaker::new(ptr, waker_vtable!(Self)))
         }
     }
 }
 
+// FIXME: panics on Arc::clone / refcount changes could wreak havoc on the
+// code here. We should guard against this by aborting.
+
 unsafe fn increase_refcount<T: ArcWake>(data: *const()) {
     // Retain Arc by creating a copy
     let arc: Arc<T> = Arc::from_raw(data as *const T);
     let arc_clone = arc.clone();
     // Forget the Arcs again, so that the refcount isn't decrased
-    let _ = Arc::into_raw(arc);
-    let _ = Arc::into_raw(arc_clone);
+    mem::forget(arc);
```

Comment:
I used `Arc::into_raw` since that's the version where the API guarantees that it's the opposite of `Arc::from_raw`.
forget will work too, but I thought if there is a more specific API available it might be nice to use it.
Anyway, both is fine.